x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

```diff
@@ -7,55 +7,204 @@
+import org.elasticsearch.xpack.esql.core.expression.Alias;
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
 import org.elasticsearch.xpack.esql.core.expression.FieldAttribute;
+import org.elasticsearch.xpack.esql.core.expression.NameId;
+import org.elasticsearch.xpack.esql.core.expression.ReferenceAttribute;
 import org.elasticsearch.xpack.esql.expression.Order;
+import org.elasticsearch.xpack.esql.expression.function.scalar.spatial.BinarySpatialFunction;
+import org.elasticsearch.xpack.esql.expression.function.scalar.spatial.SpatialRelatesUtils;
+import org.elasticsearch.xpack.esql.expression.function.scalar.spatial.StDistance;
 import org.elasticsearch.xpack.esql.optimizer.LocalPhysicalOptimizerContext;
 import org.elasticsearch.xpack.esql.optimizer.PhysicalOptimizerRules;
 import org.elasticsearch.xpack.esql.plan.physical.EsQueryExec;
+import org.elasticsearch.xpack.esql.plan.physical.EvalExec;
 import org.elasticsearch.xpack.esql.plan.physical.ExchangeExec;
 import org.elasticsearch.xpack.esql.plan.physical.PhysicalPlan;
 import org.elasticsearch.xpack.esql.plan.physical.TopNExec;
 
 import java.util.ArrayList;
+import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.function.Predicate;
 
 public class PushTopNToSource extends PhysicalOptimizerRules.ParameterizedOptimizerRule<TopNExec, LocalPhysicalOptimizerContext> {
     @Override
     protected PhysicalPlan rule(TopNExec topNExec, LocalPhysicalOptimizerContext ctx) {
-        PhysicalPlan plan = topNExec;
-        PhysicalPlan child = topNExec.child();
-        if (canPushSorts(child)
-            && canPushDownOrders(topNExec.order(), x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats()))) {
+        Pushable pushable = evaluatePushable(topNExec, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats()));
+        return pushable.rewrite(topNExec);
+    }
+
+    /**
+     * Multiple scenarios for pushing down TopN to Lucene source. Each involve checking a combination of conditions and then
+     * performing an associated rewrite specific to that scenario. This interface should be extended by each scenario, and
+     * include the appropriate rewrite logic.
+     */
+    interface Pushable {
+        PhysicalPlan rewrite(TopNExec topNExec);
+    }
+
+    private static Pushable NO_OP = new NoOpPushable();
+
+    record NoOpPushable() implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            return topNExec;
+        }
+    }
+
+    record PushableExchangeExec(ExchangeExec exchangeExec, EsQueryExec queryExec) implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            var sorts = buildFieldSorts(topNExec.order());
+            var limit = topNExec.limit();
+            return exchangeExec.replaceChild(queryExec.withSorts(sorts).withLimit(limit));
```

Comment:
Good question. This part of the code pre-existed my work, and so I kept in in place, but have to be honest I think it is surprising. Perhaps I just delete it, and see if no tests fail, and assume test coverage is sufficient? But since it pre-existed, and does not relate to my work, I decided to just leave it in place.